### ISSUE TITLE
Support file objects

### ISSUE THREAD

**Issue by mrocklin** [maxpoint/cyavro#5]
Sometimes I want to read from or write two raw bytes in memory.  In these cases I'm used to supplying a BytesIO object.

``` python
In [1]: from io import BytesIO, StringIO

In [2]: f = BytesIO(b'name,amount\nAlice,100\nBob,200')

In [3]: import pandas as pd

In [4]: df = pd.read_csv(f)

In [5]: df
Out[5]: 
    name  amount
0  Alice     100
1    Bob     200

In [6]: out = StringIO()

In [7]: df.to_csv(out)

In [8]: out.seek(0)
Out[8]: 0

In [9]: out.read()
Out[9]: ',name,amount\n0,Alice,100\n1,Bob,200\n'
```

If relevant reading and writing functions would support file objects as well as filenames then I could more easily use `cyavro` within distributed memory applications.

**Comment by mariusvniekerk** [comment]
Yeah this is something that I would like to still add.  Not entirely sure what the most effective way is to go from a file-like object to a void*

**Comment by mrocklin** [comment]
What does the `void*` point to?

**Comment by mariusvniekerk** [comment]
Basically you need to pass a pointer to a buffer housing the bytes to
libavro.

There was some recent stuff added in a pr that should be useful for this.

**Comment by mariusvniekerk** [comment]
The main issue is that libavro's C api REALLY wants files.  So the only way to really do this is by faking POSIX files

https://github.com/apache/avro/blob/trunk/lang/c/src/datafile.c#L76

**Comment by mrocklin** [comment]
The code below seems to imply that we can also pass (heavily wrapped) `char *`s to these functions:

``` c
enum avro_io_type_t {
    AVRO_FILE_IO,
    AVRO_MEMORY_IO
};
typedef enum avro_io_type_t avro_io_type_t;

struct avro_reader_t_ {
    avro_io_type_t type;
    volatile int  refcount;
};

struct avro_writer_t_ {
    avro_io_type_t type;
    volatile int  refcount;
};

struct _avro_reader_file_t {
    struct avro_reader_t_ reader;
    FILE *fp;
    int should_close;
    char *cur;
    char *end;
    char buffer[4096];
};

struct _avro_writer_file_t {
    struct avro_writer_t_ writer;
    FILE *fp;
    int should_close;
};

struct _avro_reader_memory_t {
    struct avro_reader_t_ reader;
    const char *buf;
    int64_t len;
    int64_t read;
};

struct _avro_writer_memory_t {
    struct avro_writer_t_ writer;
    const char *buf;
    int64_t len;
    int64_t written;
};
```

Also, cc-ing @kwmsmith , who might be of some assistance on the Cython side.

**Comment by mariusvniekerk** [comment]
I've finally managed to scrape away some time to take a look at the difference between how _avro_reader_file_t and _avro_reader_memory get initialized.  This seems mostly to boil down to the single avro record that is the header of the avro file that is not dealt with by the memory protocol.  (This is the way that the python avro implementation actually reads its schema).

**Comment by mariusvniekerk** [comment]
Figured out what was missing.  Implemented in #6
